### PR TITLE
CK Instance Gen

### PR DESCRIPTION
**Pull request by arai713**
no-ci

### CODE REVIEW

**Comment by arai713** [comment]
I am still working on properly including the headers

File: codegen/driver/main.cpp
Comment (bartekxk) [inline codegen/driver/main.cpp]: Should it be removed?

File: codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_op.hpp
Comment (bartekxk) [inline codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_op.hpp:29]: Maybe use multiple A and multiple B implementation? It will be more flexible.

File: codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_problem.hpp
Comment (bartekxk) [inline codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_problem.hpp]:
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.

```

File: codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_problem.hpp
Comment (bartekxk) [inline codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_problem.hpp]: Only 1d and 2d are supported?

File: codegen/include/ck/host/types.hpp
Comment (bartekxk) [inline codegen/include/ck/host/types.hpp:51]: At now we support only GNHWC/GNHWK and NHWGC/NHWGK. Codegen supports more layouts?

File: codegen/src/device_grouped_conv_fwd_multiple_d.cpp
Comment (bartekxk) [inline codegen/src/device_grouped_conv_fwd_multiple_d.cpp]: Maybe change this file name to multiple_abd?

File: codegen/test/grouped_conv_fwd_multiple_d_2.cpp
Comment (bartekxk) [inline codegen/test/grouped_conv_fwd_multiple_d_2.cpp]: Could tests for 1ds, 2ds and 3ds be templated to one test? It's large ammount of duplicated code.

File: codegen/test/grouped_conv_fwd_multiple_d_2.cpp
Comment (bartekxk) [inline codegen/test/grouped_conv_fwd_multiple_d_2.cpp]: The name of this test is a little bit confusing. Maybe use some ElementwiseOp for each of ds1,ds2,ds3.. d4 (scalar, bilinear, ConvInvscale)?

File: codegen/test/rtc/src/compile_kernel.cpp
Comment (bartekxk) [inline codegen/test/rtc/src/compile_kernel.cpp:98]: Should it be removed?

File: codegen/test/rtc/src/hip.cpp
Comment (bartekxk) [inline codegen/test/rtc/src/hip.cpp]: Should it be removed?

File: include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp]:
```suggestion
// Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```

File: include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp
Comment (bartekxk) [inline include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp:572]: Random place of the file: Could this file be just included? What is the reason to duplice them?

File: include/ck/tensor_operation/operator_transform/codegen_transform_conv_fwd_to_gemm.hpp
Comment (bartekxk) [inline include/ck/tensor_operation/operator_transform/codegen_transform_conv_fwd_to_gemm.hpp]: Random place: Why not include this file? Is some reason to duplicate them?

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (bartekxk) [inline include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp]: Can you explain? We probably should avoid relative paths in includes.

File: codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_problem.hpp
Comment (arai713) [reply codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_problem.hpp]: That's a mistake. 1d, 2d, and 3d are supported.

File: codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_op.hpp
Comment (arai713) [reply codegen/include/ck/host/device_grouped_conv_fwd_multiple_d/conv_fwd_op.hpp:29]: I have this planned as a future development, but not as part of this PR
(https://ontrack-internal.amd.com/browse/LWPCK-1424)

File: codegen/test/rtc/src/compile_kernel.cpp
Comment (arai713) [reply codegen/test/rtc/src/compile_kernel.cpp:98]: I left that in the case I need to dump the code object in ELF format. What is the convention to leave debug code?

File: codegen/test/grouped_conv_fwd_multiple_d_2.cpp
Comment (arai713) [reply codegen/test/grouped_conv_fwd_multiple_d_2.cpp]: I think my file names are a bit confusing. The different test files for grouped convolution are different convolution stride/paddings and input left/right paddings, not different dimensions.

File: codegen/test/grouped_conv_fwd_multiple_d_2.cpp
Comment (arai713) [reply codegen/test/grouped_conv_fwd_multiple_d_2.cpp]: Each of the tests use a user-provided epilogue function or PassThrough as a default

File: include/ck/tensor_operation/operator_transform/codegen_transform_conv_fwd_to_gemm.hpp
Comment (arai713) [reply include/ck/tensor_operation/operator_transform/codegen_transform_conv_fwd_to_gemm.hpp]: Will merge this file with original CK file using overloaded functions

File: include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp
Comment (arai713) [reply include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp:572]: Too difficult to reuse code due to changes in Argument class

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (bartekxk) [inline include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:19]: Maybe reuse accumulate_n from library/include/ck/library/utility/numeric.hpp ?

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (bartekxk) [inline include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:627]: Please dont duplicate code of MakeDescriptor functions. Can you just leave function body and call first one from second one?
Something like:

```
    __host__ __device__ static auto
    MakeADescriptor_M_K(const ck::Array<index_t, NDimSpatial + 3>& a_g_n_c_wis_lengths,
                        const ck::Array<index_t, NDimSpatial + 3>& a_g_n_c_wis_strides,
                        const ck::Array<index_t, NDimSpatial + 3>& b_g_k_c_xs_lengths,
                        const ck::Array<index_t, NDimSpatial + 3>& /* b_g_k_c_xs_strides */,
                        const ck::Array<index_t, NDimSpatial + 3>& c_g_n_k_wos_lengths,
                        const ck::Array<index_t, NDimSpatial + 3>& /* c_g_n_k_wos_strides */,
                        const ck::Array<index_t, NDimSpatial>& conv_filter_strides,
                        const ck::Array<index_t, NDimSpatial>& conv_filter_dilations,
                        const ck::Array<index_t, NDimSpatial>& input_left_pads,
                        const ck::Array<index_t, NDimSpatial>& input_right_pads)
    {
     /* Make descriptor logic here */
     }

    template <typename ALayout,
              typename std::enable_if<NDimSpatial == 1 &&
                                          (is_same_v<ALayout, tensor_layout::convolution::G_NW_C> ||
                                           is_same_v<ALayout, tensor_layout::convolution::NWGC> ||
                                           is_same_v<ALayout, tensor_layout::convolution::GNWC>),
                                      bool>::type = false>
    __host__ __device__ static auto
    MakeADescriptor_M_K(const ck::Array<index_t, NDimSpatial + 3>& a_g_n_c_wis_lengths,
                        const ck::Array<index_t, NDimSpatial + 3>& a_g_n_c_wis_strides,
                        const ck::Array<index_t, NDimSpatial + 3>& b_g_k_c_xs_lengths,
                        const ck::Array<index_t, NDimSpatial + 3>& /* b_g_k_c_xs_strides */,
                        const ck::Array<index_t, NDimSpatial + 3>& c_g_n_k_wos_lengths,
                        const ck::Array<index_t, NDimSpatial + 3>& /* c_g_n_k_wos_strides */,
                        const ck::Array<index_t, NDimSpatial>& conv_filter_strides,
                        const ck::Array<index_t, NDimSpatial>& conv_filter_dilations,
                        const ck::Array<index_t, NDimSpatial>& input_left_pads,
                        const ck::Array<index_t, NDimSpatial>& input_right_pads)
    {
/*Call previous one with these params */ 

}
     
```

**Review by bartekxk — COMMENTED** [review]
LGTM, just please fix these two comments

File: codegen/test/grouped_conv_fwd_multiple_d_v4.cpp
Comment (bartekxk) [inline codegen/test/grouped_conv_fwd_multiple_d_v4.cpp:200]: The same question again here:
Can we merge some part of code from these v1, v2,v3,v4 into some common?

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (arai713) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:19]: The version in CK uses std::accumulate, which I believe is not supported for device code.

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (arai713) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:627]: As I understand it, CK defines 3 versions of the functions based on the number of dimensions, so they all have a different implementation. When I overload the functions, I will need to overload all of the original functions.

Comment (bartekxk) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:627]: Yes but, what I mean is not to duplicate code for the same dimensions. You have created two identical functions but with different parameters. One of them may call the second one.

**Review by bartekxk — DISMISSED** [review]
I will take care about duplicated code in conv_fwd_transfer in post merge PR

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (arai713) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:627]: Switching std::array to ck::Array will be a PR in the future

**Comment by zjing14** [comment]
@bartekxk @tenpercent Could you approve the PR if you are OK with it?

File: codegen/driver/main.cpp
Comment (bartekxk) [inline codegen/driver/main.cpp]:
```suggestion
    // Specify problem type and problem size
```

File: codegen/driver/main.cpp
Comment (bartekxk) [inline codegen/driver/main.cpp]:
```suggestion
    // Join together all the strings in the map
```

File: codegen/driver/main.cpp
Comment (bartekxk) [inline codegen/driver/main.cpp]:
```suggestion
    // retrieve the hard-coded instances provided and template them (store in a map)
```

File: codegen/src/device_gemm_multiple_d_operation_xdl_cshuffle.cpp
Comment (bartekxk) [inline codegen/src/device_gemm_multiple_d_operation_xdl_cshuffle.cpp]:
```suggestion
// accounts for all possible combinations of Row/Col major
```

File: codegen/src/device_gemm_multiple_d_operation_xdl_cshuffle.cpp
Comment (bartekxk) [inline codegen/src/device_gemm_multiple_d_operation_xdl_cshuffle.cpp]:
```suggestion
// use hardcoded instances from vector of operations to substitute values into instance template
```

File: codegen/test/grouped_conv_fwd_multiple_d_v4.cpp
Comment (bartekxk) [inline codegen/test/grouped_conv_fwd_multiple_d_v4.cpp]:
```suggestion
        // first instance
```

**Review by bartekxk — DISMISSED** [review]
LGTM, please just fix typos

File: codegen/CMakeLists.txt
Comment (amberhassaan) [inline codegen/CMakeLists.txt]: How do the kernels depended on `utility` lib?

File: codegen/CMakeLists.txt
Comment (amberhassaan) [inline codegen/CMakeLists.txt]: minor: remove.

File: codegen/include/ck/host/device_gemm_multiple_d/problem.hpp
Comment (amberhassaan) [inline codegen/include/ck/host/device_gemm_multiple_d/problem.hpp:15]: [This can be addressed in a followup PR] : These problem structs should have problem builders that build a problem by setting various fields below. Such as: 
```
auto problem = ProblemBuilder{}.setM(m).setN(n).setK(k)...build();
```
This way, the builder controls which fields can be optional and which are required because `build` would fail if the required fields are not all specified. 

Also should be renamed to `GemmProblem` if we are naming Conv problems things like `Problem_Conv_Fwd` etc.

File: codegen/src/device_grouped_conv_fwd_multiple_abd_operation_xdl_cshuffle.cpp
Comment (amberhassaan) [inline codegen/src/device_grouped_conv_fwd_multiple_abd_operation_xdl_cshuffle.cpp]: better to replace `5` with `NumDim + 2`. and create a type alias for this `ck::Array` instantiation, e.g. `using Array5D = ck::Array<index_t, 5>; `.

File: codegen/test/rtc/src/compile_kernel.cpp
Comment (amberhassaan) [reply codegen/test/rtc/src/compile_kernel.cpp:98]: You can define a method that does this for you. Put `debug` somewhere in the name to make the intention clear.

File: include/ck/tensor_operation/gpu/device/helper.hpp
Comment (amberhassaan) [inline include/ck/tensor_operation/gpu/device/helper.hpp]: When `if` has a `return`, it is preferred not to use `else if`. Rather you should do: 
```
if (cond_a) {
   return x;
}
if (cond b) {
  return y;
}
return blah; // if all the above conds fail
```
See: https://releases.llvm.org/10.0.0/docs/CodingStandards.html#don-t-use-else-after-a-return

Benefits are readability and lesser jumps and tests.

Comment (amberhassaan) [reply include/ck/tensor_operation/gpu/device/helper.hpp]: Same applies to many of the functions in this file. You can take this up as a follow up task.

File: include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp
Comment (amberhassaan) [reply include/ck/tensor_operation/gpu/device/impl/codegen_device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp:572]: @bartekxk : Ideal solution would be to modify (instead of duplicate) all such files so that they can be used both by conventional CK and Astha's codegen.

File: include/ck/tensor_operation/gpu/device/impl/device_gemm_multiple_d_xdl_cshuffle.hpp
Comment (amberhassaan) [inline include/ck/tensor_operation/gpu/device/impl/device_gemm_multiple_d_xdl_cshuffle.hpp]: let's remove this unnecessary change.

File: include/ck/tensor_operation/gpu/device/matrix_padder.hpp
Comment (amberhassaan) [inline include/ck/tensor_operation/gpu/device/matrix_padder.hpp]: No need to define this constructor.

File: include/ck/tensor_operation/gpu/device/matrix_padder.hpp
Comment (amberhassaan) [inline include/ck/tensor_operation/gpu/device/matrix_padder.hpp]: which means this could be a static function or a standalone function outside of the struct?

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_multiple_d_xdl_cshuffle.hpp
Comment (amberhassaan) [inline include/ck/tensor_operation/gpu/grid/gridwise_gemm_multiple_d_xdl_cshuffle.hpp]: unnecessary changes should be removed.

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (amberhassaan) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:19]: Correct.

File: library/include/ck/library/utility/numeric.hpp
Comment (amberhassaan) [inline library/include/ck/library/utility/numeric.hpp]: we can undo the changes in this file as they are not needed.

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (amberhassaan) [inline include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:1156]: Seems like you don't need conditionals as you're doing the same thing in every `if/else-if`?

**Review by amberhassaan — COMMENTED** [review]
Some comments to address before merging.

File: codegen/CMakeLists.txt
Comment (amberhassaan) [reply codegen/CMakeLists.txt]: OK, I see that Array is from `utility`. Ignore the above please.

File: include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp
Comment (arai713) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:1156]: The layouts being passed in differ based on the number of dimensions

Comment (amberhassaan) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:1156]: ok, I see. In that case, one solution could be to take `Layout` as a template parameter (and function parameter). Another more immediate solution could be to use `if constexpr` in all conditionals so it's resolved at compile time.

Comment (arai713) [reply include/ck/tensor_operation/operator_transform/transform_conv_fwd_to_gemm.hpp:1156]: It's difficult to pass in layout as a template argument, since it is usually a string from the generated code and will be different based on the number of dimensions and won't be a constant expression. Additionally, for passing as a function parameter, each layout is defined as its own struct, so that becomes difficult as well. All the arguments used in the launch parameter calculation are found at runtime, so I don't think they can be resolved at compile time

File: codegen/include/ck/host/device_gemm_multiple_d/problem.hpp
Comment (arai713) [reply codegen/include/ck/host/device_gemm_multiple_d/problem.hpp:15]: I will add this in a subsequent PR

**Review by zjing14 — APPROVED** [review]

**Pull request merged**
